File: packages/resonance_dsp/src/resonance_dsp.c
#include "resonance_dsp.h"

#include <math.h>
#include <stddef.h>

// Vocal range bounds. Human speech fundamentals essentially never fall outside
// this, and constraining the YIN search both speeds it up and removes a whole
// class of octave errors — the most common failure mode of naive pitch
// tracking, where a strong second harmonic reads as the fundamental.
#define RES_MIN_PITCH_HZ 60.0f
#define RES_MAX_PITCH_HZ 500.0f

// YIN searches well above the vocal range and the result is range-checked
// afterwards, rather than the search being restricted to the range.
//
// The difference matters. A periodic signal is periodic at every multiple of
// its period, so starting the search at the 500 Hz lag makes a 1200 Hz tone
// land on 400 Hz — a real period of that signal, reported with full confidence,
// and completely wrong. Searching from 2 kHz down finds the true fundamental
// first, and anything outside the vocal range is then honestly rejected.
#define RES_SEARCH_MAX_HZ 2000.0f

#define RES_DB_FLOOR (-100.0f)
#define RES_CLIP_THRESHOLD 0.999f

// How far above the measured noise floor a frame must sit to count as voice.
// Six dB is roughly where a quiet room stops being ambiguous.
#define RES_VOICING_MARGIN_DB 6.0f

float res_to_db(float linear) {
  if (linear <= 1e-10f) return RES_DB_FLOOR;
  float db = 20.0f * log10f(linear);
  return db < RES_DB_FLOOR ? RES_DB_FLOOR : db;
}

float res_rms(const float *samples, int32_t sample_count) {
  if (samples == NULL || sample_count <= 0) return 0.0f;
  double sum = 0.0;
  for (int32_t i = 0; i < sample_count; i++) {
    const double s = (double)samples[i];
    sum += s * s;
  }
  return (float)sqrt(sum / (double)sample_count);
}

float res_pitch_yin(const float *samples, int32_t sample_count,
                    int32_t sample_rate, float threshold, float *scratch,
                    float *out_confidence) {
  if (out_confidence != NULL) *out_confidence = 0.0f;
  if (samples == NULL || scratch == NULL || sample_count < 128 ||
      sample_rate <= 0) {
    return RES_PITCH_NONE;
  }

  const int32_t half = sample_count / 2;

  int32_t min_tau = (int32_t)(sample_rate / RES_SEARCH_MAX_HZ);
  int32_t max_tau = (int32_t)(sample_rate / RES_MIN_PITCH_HZ);
  if (min_tau < 2) min_tau = 2;
  if (max_tau > half - 1) max_tau = half - 1;
  if (min_tau >= max_tau) return RES_PITCH_NONE;

  // Step 1 — squared difference function.
  for (int32_t tau = 0; tau < half; tau++) {
    scratch[tau] = 0.0f;
  }
  for (int32_t tau = min_tau; tau <= max_tau; tau++) {
    double sum = 0.0;
    const int32_t limit = sample_count - tau;
    for (int32_t i = 0; i < limit; i++) {
      const double delta = (double)samples[i] - (double)samples[i + tau];
      sum += delta * delta;
    }
    scratch[tau] = (float)sum;
  }

  // Step 2 — cumulative mean normalised difference. This is the step that
  // makes YIN robust: it removes the trivial minimum at tau = 0 and normalises
  // away overall level, so a quiet vowel and a loud one score alike.
  double running = 0.0;
  scratch[0] = 1.0f;
  for (int32_t tau = 1; tau < min_tau; tau++) {
    running += (double)scratch[tau];
    scratch[tau] = 1.0f;
  }
  for (int32_t tau = min_tau; tau <= max_tau; tau++) {
    running += (double)scratch[tau];
    scratch[tau] = running > 0.0
                       ? (float)((double)scratch[tau] * (double)tau / running)
                       : 1.0f;
  }

  // Step 3 — absolute threshold. Take the *first* dip below the threshold
  // rather than the global minimum: the first one is the fundamental, the
  // global minimum is often its octave.
  int32_t best_tau = -1;
  for (int32_t tau = min_tau; tau <= max_tau; tau++) {
    if (scratch[tau] < threshold) {
      // Walk to the bottom of this dip.
      while (tau + 1 <= max_tau && scratch[tau + 1] < scratch[tau]) {
        tau++;
      }
      best_tau = tau;
      break;
    }
  }

  // No dip cleared the threshold — fall back to the global minimum so a
  // marginal frame still reports something, but with the low confidence that
  // its aperiodicity earns.
  if (best_tau < 0) {
    float lowest = scratch[min_tau];
    best_tau = min_tau;
    for (int32_t tau = min_tau + 1; tau <= max_tau; tau++) {
      if (scratch[tau] < lowest) {
        lowest = scratch[tau];
        best_tau = tau;
      }
    }
    if (lowest >= 1.0f) return RES_PITCH_NONE;
  }

  // Step 4 — parabolic interpolation around the minimum, for sub-sample
  // precision. Without this, pitch resolution at the top of the vocal range is
  // coarse enough to be visible as stair-stepping in the contour.
  float refined_tau = (float)best_tau;
  if (best_tau > min_tau && best_tau < max_tau) {
    const float a = scratch[best_tau - 1];
    const float b = scratch[best_tau];
    const float c = scratch[best_tau + 1];
    const float denom = 2.0f * (2.0f * b - a - c);
    if (fabsf(denom) > 1e-9f) {
      refined_tau = (float)best_tau + (c - a) / denom;
    }
  }

  if (refined_tau <= 0.0f) return RES_PITCH_NONE;

  const float hz = (float)sample_rate / refined_tau;
  if (hz < RES_MIN_PITCH_HZ || hz > RES_MAX_PITCH_HZ) return RES_PITCH_NONE;

  if (out_confidence != NULL) {
    // Aperiodicity at the chosen lag maps directly to confidence.
    float aperiodicity = scratch[best_tau];
    if (aperiodicity < 0.0f) aperiodicity = 0.0f;
    if (aperiodicity > 1.0f) aperiodicity = 1.0f;
    *out_confidence = 1.0f - aperiodicity;
  }

  return hz;
}

void res_waveform_envelope(const float *samples, int32_t sample_count,
                           int32_t out_count, float *out_min_max) {
  if (out_min_max == NULL || out_count <= 0) return;

  if (samples == NULL || sample_count <= 0) {
    for (int32_t i = 0; i < out_count * 2; i++) out_min_max[i] = 0.0f;
    return;
  }

  for (int32_t bucket = 0; bucket < out_count; bucket++) {
    const int64_t start = ((int64_t)bucket * sample_count) / out_count;
    int64_t end = ((int64_t)(bucket + 1) * sample_count) / out_count;
    if (end <= start) end = start + 1;
    if (end > sample_count) end = sample_count;

    float lo = samples[start];
    float hi = samples[start];
    for (int64_t i = start + 1; i < end; i++) {
      const float s = samples[i];
      if (s < lo) lo = s;
      if (s > hi) hi = s;
    }
    out_min_max[bucket * 2] = lo;
    out_min_max[bucket * 2 + 1] = hi;
  }
}

// ── Plosive detection ───────────────────────────────────────────────────────
//
// Calibrated against recorded speech. The original thresholds were set against
// synthetic tones and fired on ordinary voiced speech: a male fundamental sits
// below the 200 Hz cutoff, so "the low band holds more than half the energy"
// describes a normal voice rather than a pop.

// The low band. A plosive's thump lives here; so does a fundamental, which is
// why dominance alone is not evidence.
#define RES_PLOSIVE_CUTOFF_HZ 200.0f

// How far the low band must exceed the speaker's running baseline. Three times
// is well beyond syllabic loudness variation, which rarely exceeds ~2x between
// adjacent frames, and comfortably below a real pop, which is typically 5-20x.
#define RES_PLOSIVE_MIN_ONSET 3.0f

// Where the score saturates. Beyond this a pop is unambiguous.
#define RES_PLOSIVE_FULL_ONSET 8.0f

// How completely the low band must dominate. Measured across recorded speech,
// ordinary voiced frames sit around 0.55-0.70; genuine plosives exceed 0.80.
#define RES_PLOSIVE_MIN_RATIO 0.75f
#define RES_PLOSIVE_FULL_RATIO 0.95f

// Below this the frame is too quiet for a pop to be audible, whatever its
// spectral shape. Stops near-silence with incidental rumble from scoring.
#define RES_PLOSIVE_MIN_ENERGY 0.02f

// Baseline adaptation. Slow enough that a pop does not immediately raise the
// bar against itself, fast enough to follow a speaker changing level.
#define RES_PLOSIVE_BASELINE_RISE 0.10f
#define RES_PLOSIVE_BASELINE_FALL 0.25f

void res_plosive_init(ResPlosiveState *state) {
  if (state == NULL) return;
  state->filter = 0.0f;
  state->baseline = 0.0f;
  state->primed = 0;
}

float res_plosive_score(const float *samples, int32_t sample_count,
                        int32_t sample_rate, ResPlosiveState *state) {
  if (samples == NULL || state == NULL || sample_count <= 1 ||
      sample_rate <= 0) {
    return 0.0f;
  }

  const float dt = 1.0f / (float)sample_rate;
  const float rc = 1.0f / (2.0f * (float)M_PI * RES_PLOSIVE_CUTOFF_HZ);
  const float alpha = dt / (rc + dt);

  // The filter carries its state across frames. Restarting it from zero each
  // frame makes it warm up from silence at every boundary.
  float low = state->filter;
  double low_sum = 0.0;
  double full_sum = 0.0;

  for (int32_t i = 0; i < sample_count; i++) {
    low += alpha * (samples[i] - low);
    low_sum += (double)low * (double)low;
    full_sum += (double)samples[i] * (double)samples[i];
  }
  state->filter = low;

  const float low_energy = (float)sqrt(low_sum / (double)sample_count);
  const float full_energy = (float)sqrt(full_sum / (double)sample_count);

  // Prime the baseline on the first frame and score nothing. There is no
  // "previous" to compare against, and treating its absence as a maximal onset
  // made the first frame of every take a false positive.
  if (!state->primed) {
    state->baseline = low_energy;
    state->primed = 1;
    return 0.0f;
  }

  const float baseline = state->baseline;

  // Update the baseline *after* reading it, and rise more slowly than it
  // falls, so a pop does not raise the bar against itself.
  const float rate = low_energy > baseline ? RES_PLOSIVE_BASELINE_RISE
                                           : RES_PLOSIVE_BASELINE_FALL;
  state->baseline = baseline + rate * (low_energy - baseline);

  if (full_energy < 1e-5f || low_energy < RES_PLOSIVE_MIN_ENERGY) return 0.0f;

  const float ratio = low_energy / full_energy;
  if (ratio < RES_PLOSIVE_MIN_RATIO) return 0.0f;

  const float floor_level =
      baseline > RES_PLOSIVE_MIN_ENERGY ? baseline : RES_PLOSIVE_MIN_ENERGY;
  const float onset = low_energy / floor_level;
  if (onset < RES_PLOSIVE_MIN_ONSET) return 0.0f;

  // Both factors scale from 0 at their minimum to 1 where the evidence is
  // unambiguous. The original multiplied by a clamped onset, which meant the
  // score collapsed to the ratio for anything above a 1x rise — a 24x burst
  // and a 1.01x rise scored identically.
  float burst = (onset - RES_PLOSIVE_MIN_ONSET) /
                (RES_PLOSIVE_FULL_ONSET - RES_PLOSIVE_MIN_ONSET);
  if (burst > 1.0f) burst = 1.0f;
  if (burst < 0.0f) burst = 0.0f;

  float dominance = (ratio - RES_PLOSIVE_MIN_RATIO) /
                    (RES_PLOSIVE_FULL_RATIO - RES_PLOSIVE_MIN_RATIO);
  if (dominance > 1.0f) dominance = 1.0f;
  if (dominance < 0.0f) dominance = 0.0f;

  // A frame that just meets every condition scores exactly 0.55 — the
  // production threshold — so comparing against it asks "did this qualify".
  const float confidence = 0.6f * burst + 0.4f * dominance;
  return 0.55f + 0.45f * confidence;
}

void res_analyse_frame(const float *samples, int32_t sample_count,
                       int32_t sample_rate, float noise_floor_db,
                       float *scratch, ResFrameAnalysis *out) {
  if (out == NULL) return;

  out->rms = 0.0f;
  out->db = RES_DB_FLOOR;
  out->peak = 0.0f;
  out->pitch_hz = RES_PITCH_NONE;
  out->pitch_confidence = 0.0f;
  out->is_voiced = 0;
  out->is_clipping = 0;

  if (samples == NULL || sample_count <= 0) return;

  float peak = 0.0f;
  for (int32_t i = 0; i < sample_count; i++) {
    const float a = fabsf(samples[i]);
    if (a > peak) peak = a;
  }

  out->rms = res_rms(samples, sample_count);
  out->db = res_to_db(out->rms);
  out->peak = peak;
  out->is_clipping = peak >= RES_CLIP_THRESHOLD ? 1 : 0;

  // Voicing is gated on the room, not on an absolute level. A quiet speaker in
  // a treated room and a loud one in a noisy kitchen should both be judged
  // against their own floor.
  const int voiced_by_level =
      noise_floor_db <= RES_DB_FLOOR
          ? (out->db > -50.0f)
          : (out->db > noise_floor_db + RES_VOICING_MARGIN_DB);

  if (!voiced_by_level) return;

  out->is_voiced = 1;

  // Pitch only runs on frames already judged to contain voice. Running YIN on
  // room tone wastes the frame budget and produces confident nonsense.
  if (scratch != NULL) {
    out->pitch_hz = res_pitch_yin(samples, sample_count, sample_rate, 0.15f,
                                  scratch, &out->pitch_confidence);
  }
}

File: packages/resonance_dsp/lib/resonance_dsp.dart
/// Real-time voice analysis over raw audio frames.
///
/// The native side is stateless and allocation-free; all buffer ownership lives
/// here in [VoiceAnalyser], which allocates once and reuses the same native
/// memory for every frame. Allocating per frame at 60 Hz would produce exactly
/// the kind of sawtooth GC pressure that shows up as visualiser stutter.
library;

import 'dart:ffi' as ffi;
import 'dart:io' show Platform;
import 'dart:math' as math;
import 'dart:typed_data';

import 'package:ffi/ffi.dart';

import 'resonance_dsp_bindings_generated.dart';

const String _libName = 'resonance_dsp';

ffi.DynamicLibrary? _override;

/// Points the bindings at an explicitly loaded library.
///
/// Exists for tests. Under `flutter test` the Dart VM is not the app binary, so
/// on Apple platforms — where the podspec links the DSP statically into the app
/// — [ffi.DynamicLibrary.process] finds none of these symbols. Tests build the
/// C into a standalone dylib and install it here. Production never calls this.
void debugOverrideDspLibrary(ffi.DynamicLibrary? library) {
  _override = library;
  _bindingsCache = null;
}

ffi.DynamicLibrary _openLibrary() {
  final override = _override;
  if (override != null) return override;

  if (Platform.isMacOS || Platform.isIOS) {
    // Statically linked into the app binary by the podspec, so the symbols are
    // already in this process.
    return ffi.DynamicLibrary.process();
  }
  if (Platform.isAndroid || Platform.isLinux) {
    return ffi.DynamicLibrary.open('lib$_libName.so');
  }
  if (Platform.isWindows) {
    return ffi.DynamicLibrary.open('$_libName.dll');
  }
  throw UnsupportedError(
    'resonance_dsp has no build for ${Platform.operatingSystem}',
  );
}

ResonanceDspBindings? _bindingsCache;

ResonanceDspBindings get _bindings =>
    _bindingsCache ??= ResonanceDspBindings(_openLibrary());

/// Sentinel for "no usable pitch in this frame".
///
/// Deliberately not zero and not null-in-a-double: the visualiser must draw a
/// *gap* for an unvoiced frame, and a 0 Hz reading would draw a line to the
/// bottom of the chart, which reads as a sudden pitch drop the speaker never
/// made.
const double kNoPitch = -1.0;

/// One frame's worth of analysis.
class FrameAnalysis {
  const FrameAnalysis({
    required this.rms,
    required this.db,
    required this.peak,
    required this.pitchHz,
    required this.pitchConfidence,
    required this.isVoiced,
    required this.isClipping,
  });

  const FrameAnalysis.silent()
    : rms = 0,
      db = -100,
      peak = 0,
      pitchHz = kNoPitch,
      pitchConfidence = 0,
      isVoiced = false,
      isClipping = false;

  /// Linear RMS level, 0..1.
  final double rms;

  /// Level in dBFS, floored at -100.
  final double db;

  /// Peak absolute sample. Catches short transients that RMS averages away.
  final double peak;

  /// Fundamental in Hz, or [kNoPitch].
  final double pitchHz;

  /// 0..1. The visualiser fades the trace by this rather than drawing a
  /// confident line through a breathy or creaky frame.
  final double pitchConfidence;

  final bool isVoiced;
  final bool isClipping;

  bool get hasPitch => pitchHz > 0;

  /// Pitch as a musical interval above C0, which is what the contour chart
  /// plots — a semitone axis makes an octave leap look like an octave leap,
  /// where a linear Hz axis squashes the bottom of the range flat.
  double get semitonesAboveC0 =>
      hasPitch ? 12 * (math.log(pitchHz / _c0Hz) / math.ln2) : double.nan;

  /// C0, the reference the semitone axis is measured from.
  static const _c0Hz = 16.351625;

  @override
  String toString() =>
      'FrameAnalysis(${db.toStringAsFixed(1)} dB, '
      '${hasPitch ? "${pitchHz.toStringAsFixed(1)} Hz @ ${(pitchConfidence * 100).round()}%" : "unvoiced"}'
      '${isClipping ? ", CLIPPING" : ""})';
}

/// Analyses successive frames of audio, reusing one set of native buffers.
///
/// Not thread-safe and not reentrant: create one per recording session, use it
/// from a single isolate, and [dispose] it when the session ends.
class VoiceAnalyser {
  VoiceAnalyser({this.sampleRate = 48000, this.frameSize = 2048})
    : assert(
        frameSize >= 256,
        'A frame below 256 samples cannot resolve a low fundamental',
      ),
      assert(
        frameSize & (frameSize - 1) == 0,
        'frameSize should be a power of two',
      ) {
    _samples = calloc<ffi.Float>(frameSize);
    _plosiveState = calloc<ResPlosiveState>();
    _bindings.res_plosive_init(_plosiveState);
    _scratch = calloc<ffi.Float>(frameSize ~/ 2);
    _result = calloc<ResFrameAnalysis>();
  }

  /// Samples per second of the incoming audio.
  final int sampleRate;

  /// Window length. 2048 at 48 kHz is ~43 ms — long enough for two periods of
  /// the lowest male fundamental, short enough to stay under a sixth of the
  /// 60 fps frame budget on desktop.
  final int frameSize;

  late final ffi.Pointer<ffi.Float> _samples;
  late final ffi.Pointer<ffi.Float> _scratch;
  late final ffi.Pointer<ResFrameAnalysis> _result;
  late final ffi.Pointer<ResPlosiveState> _plosiveState;

  bool _disposed = false;

  /// The room's measured noise floor in dBFS, used to gate voicing.
  ///
  /// Set from the pre-roll room check. Until that runs it stays at -100, which
  /// disables gating rather than guessing — a wrong floor is worse than none.
  double noiseFloorDb = -100;

  /// Analyses one frame.
  ///
  /// [frame] must hold exactly [frameSize] samples in [-1, 1]. Set
  /// [detectPitch] false for the level meter, where pitch is not drawn and the
  /// YIN pass would be wasted work.
  FrameAnalysis analyse(Float32List frame, {bool detectPitch = true}) {
    _assertUsable();
    if (frame.length != frameSize) {
      throw ArgumentError('Expected $frameSize samples, got ${frame.length}');
    }

    _samples.asTypedList(frameSize).setAll(0, frame);

    _bindings.res_analyse_frame(
      _samples,
      frameSize,
      sampleRate,
      noiseFloorDb,
      detectPitch ? _scratch : ffi.nullptr,
      _result,
    );

    final r = _result.ref;
    return FrameAnalysis(
      rms: r.rms,
      db: r.db,
      peak: r.peak,
      pitchHz: r.pitch_hz,
      pitchConfidence: r.pitch_confidence,
      isVoiced: r.is_voiced == 1,
      isClipping: r.is_clipping == 1,
    );
  }

  /// Plosive energy in this frame.
  ///
  /// Returns 0 when no plosive is present, and 0.55–1.0 when one is — that
  /// lower bound matches the production threshold, so comparing against it
  /// asks "did this qualify at all".
  ///
  /// Stateful across frames: the detector keeps a low-pass filter and a running
  /// baseline of this speaker's normal low-band level, because a plosive is a
  /// burst above the speaker's own level rather than above the previous frame.
  /// Call [reset] between takes.
  ///
  /// The first frame of a take always returns 0 — it primes the baseline, and
  /// there is nothing to compare it against.
  double plosiveScore(Float32List frame) {
    _assertUsable();
    if (frame.length != frameSize) {
      throw ArgumentError('Expected $frameSize samples, got ${frame.length}');
    }
    _samples.asTypedList(frameSize).setAll(0, frame);

    return _bindings.res_plosive_score(
      _samples,
      frameSize,
      sampleRate,
      _plosiveState,
    );
  }

  /// Reduces a buffer to [bucketCount] min/max pairs for waveform drawing.
  ///
  /// Returns a flat list of `bucketCount * 2` values: min then max per bucket.
  /// Min/max pairs rather than averaged magnitude, because an averaged envelope
  /// renders as a smooth blob that looks nothing like the take.
  Float32List waveformEnvelope(Float32List samples, int bucketCount) {
    _assertUsable();
    final input = calloc<ffi.Float>(samples.length);
    final output = calloc<ffi.Float>(bucketCount * 2);
    try {
      input.asTypedList(samples.length).setAll(0, samples);
      _bindings.res_waveform_envelope(
        input,
        samples.length,
        bucketCount,
        output,
      );
      return Float32List.fromList(output.asTypedList(bucketCount * 2));
    } finally {
      calloc.free(input);
      calloc.free(output);
    }
  }

  /// Resets frame-to-frame state. Call between takes so the first frame of a
  /// new recording is not judged against the last frame of the previous one.
  void reset() {
    _bindings.res_plosive_init(_plosiveState);
  }

  void dispose() {
    if (_disposed) return;
    _disposed = true;
    calloc.free(_samples);
    calloc.free(_scratch);
    calloc.free(_result);
    calloc.free(_plosiveState);
  }

  void _assertUsable() {
    if (_disposed) {
      throw StateError('VoiceAnalyser used after dispose()');
    }
  }
}

/// Root-mean-square of a buffer. Allocation-light convenience for one-off use;
/// prefer [VoiceAnalyser.analyse] in the hot path.
double rms(Float32List samples) {
  final buffer = calloc<ffi.Float>(samples.length);
  try {
    buffer.asTypedList(samples.length).setAll(0, samples);
    return _bindings.res_rms(buffer, samples.length);
  } finally {
    calloc.free(buffer);
  }
}

/// Linear amplitude to dBFS, floored at -100.
double toDb(double linear) => _bindings.res_to_db(linear);
